File: conn/ado/TrafodionProvider/Trafodion.Data/Network/Structures/OutConnectionContext.cs
/**********************************************************************
// @@@ START COPYRIGHT @@@
//
// (C) Copyright 2009-2015 Hewlett-Packard Development Company, L.P.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//
// @@@ END COPYRIGHT @@@
********************************************************************/

using System;

namespace Trafodion.Data
{
    /// <summary>
    /// Option flags which indicate server settings or additional information contained in the message
    /// </summary>
    [Flags]
    internal enum OutContextOptions1 : uint
    {
	    IgnoreCancel = 1073741824, // (2^30)
	    ExtraOptions = 2147483648, // (2^31)
	    DownloadCertificate = 536870912 //(2^29)
    }

    /// <summary>
    /// A set of connection properties returned by the server after login.
    /// </summary>
    internal class OutConnectionContext: INetworkReply
    {
        public Version [] serverVersion;
        public short nodeId;
        public uint processId;

        public string computerName;
        public string catalog;
        public string schema;

        public OutContextOptions1 optionFlags1;
        public OutContextOptions1 optionFlags2;

        public String roleName;
        public Boolean ignoreCancel;

        public byte[] certificate;

        public void ReadFromDataStream(DataStream ds, TrafodionDBEncoder enc)
        {
            serverVersion = Version.ReadListFromDataStream(ds, enc);
            nodeId = ds.ReadInt16();
            processId = ds.ReadUInt32();

            computerName = enc.GetString(ds.ReadString(), enc.Transport);
            catalog = enc.GetString(ds.ReadString(), enc.Transport);
            schema = enc.GetString(ds.ReadString(), enc.Transport);

            optionFlags1 = (OutContextOptions1) ds.ReadUInt32();
            optionFlags2 = (OutContextOptions1) ds.ReadUInt32();

            ignoreCancel = (optionFlags1 & OutContextOptions1.IgnoreCancel) > 0;
            if ((optionFlags1 & OutContextOptions1.DownloadCertificate) > 0)
            {
                int len = ds.ReadInt32();
                certificate = ds.ReadBytes(len);
            }
            else if ((optionFlags1 & OutContextOptions1.ExtraOptions) > 0)
            {
                DecodeExtraOptions(enc.GetString(ds.ReadString(), enc.Transport));
            }
        }

        /// <summary>
        /// Helper method to extract extra options which are in string form.
        /// </summary>
        /// <param name="options"></param>
        private void DecodeExtraOptions(String options)
        {
            String[] opts = options.Split(new char[] { ';' });
            String token;
            String value;
            int index;

            for (int i = 0; i < opts.Length; i++)
            {
                opts[i] = opts[i].Trim();

                if (opts[i].Length > 0)
                {
                    index = opts[i].IndexOf('=');
                    token = opts[i].Substring(0, index).ToUpper();
                    value = opts[i].Substring(index + 1);

                    if (token.Equals("RN"))
                    {
                        this.roleName = value;
                    }
                }
            }
        }
    }
}

File: dcs/src/main/java/org/trafodion/dcs/servermt/serverDriverInputOutput/OutConnectionContext.java
/**
* @@@ START COPYRIGHT @@@

* Licensed to the Apache Software Foundation (ASF) under one
* or more contributor license agreements.  See the NOTICE file
* distributed with this work for additional information
* regarding copyright ownership.  The ASF licenses this file
* to you under the Apache License, Version 2.0 (the
* "License"); you may not use this file except in compliance
* with the License.  You may obtain a copy of the License at

*   http://www.apache.org/licenses/LICENSE-2.0

* Unless required by applicable law or agreed to in writing,
* software distributed under the License is distributed on an
* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
* KIND, either express or implied.  See the License for the
* specific language governing permissions and limitations
* under the License.

* @@@ END COPYRIGHT @@@
 */
package org.trafodion.dcs.servermt.serverDriverInputOutput;

import java.net.*;
import java.io.*;
import java.nio.*;
import java.nio.channels.*;
import java.nio.channels.spi.*;
import java.sql.SQLException;

import org.trafodion.dcs.Constants;
import org.trafodion.dcs.util.*;
import org.trafodion.dcs.servermt.ServerConstants;
import org.trafodion.dcs.servermt.ServerUtils;
import org.trafodion.dcs.servermt.serverHandler.*;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

public class OutConnectionContext {
    private static  final Log LOG = LogFactory.getLog(OutConnectionContext.class);
    
    private VersionList versionList;
    private short nodeId = 0;
    private int processId = 0;
    private String computerName = "";
    private String catalog = "";
    private String schema = "";
    private long optionFlags1 = 0;
    private long optionFlags2 = 0;
    private String _roleName = "";
    private boolean _enforceISO = false;
    private boolean _ignoreCancel = false;
    private byte [] cert = null;
    
    public OutConnectionContext(byte[] cert){
        optionFlags1 = optionFlags1 | ServerConstants.OUTCONTEXT_OPT1_DOWNLOAD_CERTIFICATE;
        this.cert = cert;
        versionList = new VersionList();
    }
    public void extractFromByteBuffer(ByteBuffer bbBuf) throws UnsupportedEncodingException {
        versionList.extractFromByteBuffer(bbBuf);

        nodeId = bbBuf.getShort();
        processId = bbBuf.getInt();
        computerName = ByteBufferUtils.extractString(bbBuf);

        catalog = ByteBufferUtils.extractString(bbBuf);
        schema = ByteBufferUtils.extractString(bbBuf);

        optionFlags1 = bbBuf.getInt();
        optionFlags2 = bbBuf.getInt();

        this._enforceISO = (optionFlags1 & ServerConstants.OUTCONTEXT_OPT1_ENFORCE_ISO88591) > 0;
        this._ignoreCancel = (optionFlags1 & ServerConstants.OUTCONTEXT_OPT1_IGNORE_SQLCANCEL) > 0;
        if((optionFlags1 & ServerConstants.OUTCONTEXT_OPT1_DOWNLOAD_CERTIFICATE) > 0) {
            cert = ByteBufferUtils.extractByteArray(bbBuf);
        }
        else if ((optionFlags1 & ServerConstants.OUTCONTEXT_OPT1_EXTRA_OPTIONS) > 0) {
            try {
                this.decodeExtraOptions(ByteBufferUtils.extractString(bbBuf));
            } catch (UnsupportedEncodingException ue) {
                LOG.error("An error occured parsing OutConnectionContext: " + ue.getMessage());
                throw ue;
            }
        }
    }
    public void decodeExtraOptions(String options) {
        String[] opts = options.split(";");
        String token;
        String value;
        int index;

        for (int i = 0; i < opts.length; i++) {
            index = opts[i].indexOf('=');
            token = opts[i].substring(0, index).toUpperCase();
            value = opts[i].substring(index + 1);

            if (token.equals("RN")) {
                this._roleName = value;
            }
        }
    }
    public void insertIntoByteBuffer(ByteBuffer bbBuf) throws UnsupportedEncodingException {
        
        versionList.insertIntoByteBuffer(bbBuf);
        bbBuf.putShort(nodeId);
        bbBuf.putInt(processId);
        ByteBufferUtils.insertString(computerName, bbBuf);

        ByteBufferUtils.insertString(catalog, bbBuf);
        ByteBufferUtils.insertString(schema, bbBuf);

        ByteBufferUtils.insertUInt(optionFlags1,bbBuf);
        ByteBufferUtils.insertUInt(optionFlags2,bbBuf);
        
        ByteBufferUtils.insertByteArray(cert, bbBuf);
    }
    public int lengthOfData() {

        int dataLength = 0;
        dataLength += versionList.lengthOfData();                                       //versionList

        dataLength += ServerConstants.SHORT_FIELD_SIZE;                                       //nodeId
        dataLength += ServerConstants.INT_FIELD_SIZE;                                         //processId
        dataLength += ByteBufferUtils.lengthOfString(computerName);                        //computerName
        dataLength += ByteBufferUtils.lengthOfString(catalog);                             //catalog
        dataLength += ByteBufferUtils.lengthOfString(schema);                              //schema
        dataLength += ServerConstants.INT_FIELD_SIZE;                                         //optionFlags1
        dataLength += ServerConstants.INT_FIELD_SIZE;                                         //optionFlags2
        dataLength += ByteBufferUtils.lengthOfByteArray(cert);                              //certificate
        return dataLength;
    }
    public VersionList getVersionList(){
        return versionList;
    }
    public void setNodeId(short nodeId){
        this.nodeId = nodeId;
    }
    public void setProcessId(int processId){
        this.processId = processId;
    }
    public void setComputerName(String computerName){
        this.computerName = computerName;        
    }
    public void setCatalog(String catalog){
        this.catalog = catalog;        
    }
    public void setSchema(String schema){
        this.schema = schema;        
    }
    public void setOptionFlags1(long optionFlags1){
        this.optionFlags1 = optionFlags1;        
    }
    public void setOptionFlags2(long optionFlags2){
        this.optionFlags2 = optionFlags2;        
    }
    public void setRoleName(String _roleName){
        this._roleName = _roleName;        
    }
}
